manager/integration/tests/test_settings.py

```diff
@@ -277,48 +308,46 @@ def test_setting_toleration_extra(core_api, apps_api):  # NOQA
 
 def wait_for_toleration_update(core_api, apps_api, count,  # NOQA
                                expected_tolerations,
-                               chk_removed_tolerations=[]):
-    not_managed_apps = [
-        "csi-attacher",
-        "csi-provisioner",
-        "csi-resizer",
-        "csi-snapshotter",
-        "longhorn-csi-plugin",
+                               chk_removed_tolerations=[],
+                               backing_images=[]):
+    skip_updates = [
         "longhorn-driver-deployer",
         "longhorn-manager",
         "longhorn-ui",
     ]
-    updated = False
+    ok = False
     for _ in range(RETRY_COUNTS):
         time.sleep(RETRY_INTERVAL_LONG)
 
-        updated = True
-        if not check_workload_update(core_api, apps_api, count):
-            updated = False
+        ok = False
+        if not check_longhorn_managed_workloads(core_api, apps_api, count,
+                                                backing_images):
             continue
 
-        pod_list = core_api.list_namespaced_pod(LONGHORN_NAMESPACE).items
+        if len(backing_images) != 0:
+            pod_list = common.list_backing_image_manager_pods(core_api)
+        else:
+            pod_list = core_api.list_namespaced_pod(LONGHORN_NAMESPACE).items
+
         for p in pod_list:
-            managed_by = p.metadata.labels.get('longhorn.io/managed-by', '')
-            if str(managed_by) != "longhorn-manager":
-                continue
-            else:
-                app_name = str(p.metadata.labels.get('app', ''))
-                assert app_name not in not_managed_apps
-
-            if p.status.phase != "Running" \
-                or not check_tolerations_set(p.spec.tolerations,
-                                             expected_tolerations,
-                                             chk_removed_tolerations):
-                updated = False
-                break
-        if updated:
+            if len(backing_images) != 0:
+                assert p.metadata.name.startswith("backing-image-manager")
+
+            n_update = check_tolerations_set_update(p.spec.tolerations,
+                                                    expected_tolerations,
+                                                    chk_removed_tolerations)
+            app_name = str(p.metadata.labels.get('app', ''))
+            if app_name in skip_updates and n_update == 0:
+                ok = True
+            elif n_update == len(expected_tolerations):
+                ok = True
+        if ok:
```

Comment:
I don't understand this part: 
1. Why doesn't this function check other longhorn-managed pods if there are backing images? 
2. Some of them don't contain label `app`. e.g., engine image pods,

Reply:
> 1. Why doesn't this function check other longhorn-managed pods if there are backing images?

Since we've decided to have a separate test for backing image I feel like it's a bit duplicated since testing of other pods already covered in test_setting_toleration.

> 2. Some of them don't contain label `app`. e.g., engine image pods,

Ones with no `app` label will be checked. Are there pods we would like to skip with no `app` label?

Reply:
Please check all components. We don't mind duplicate checks in integration tests.

The logic looks weird. As long as some of the pods update the tolerations, flag `ok` will be true. It means this function won't wait for all pod update complete. Here, the for loop can be:
```
ok = True
for p in pod_list:
    if p in update_skip_list:
        continue
    if not check_tolerations_set_update():
        ok = False
    if not ok:
        break

```

BTW, the field name `n_update` is confusing as well...